google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java

```diff
@@ -2840,61 +2840,75 @@ private Status processResults(
-          com.google.spanner.v1.StructType.Field.newBuilder()
-              .setName(result.getType().getStructFields().get(i).getName())
-              .setType(cloudTypeToTypeProto(columnType))
-              .build());
+              com.google.spanner.v1.StructType.Field.newBuilder()
+                      .setName(struct.getType().getStructFields().get(i).getName())
+                      .setType(cloudTypeToTypeProto(columnType))
+                      .build());
+    }
+    return rowTypeBuilder.build();
+  }
+
+  /** Convert a struct to a proto(value list) for constructing result rows and struct values. */
+  private com.google.spanner.executor.v1.ValueList buildStruct(StructReader struct) {
+    com.google.spanner.executor.v1.ValueList.Builder structBuilder =
+        com.google.spanner.executor.v1.ValueList.newBuilder();
+    for (int i = 0; i < struct.getColumnCount(); ++i) {
+      com.google.cloud.spanner.Type columnType = struct.getColumnType(i);
       com.google.spanner.executor.v1.Value.Builder value =
           com.google.spanner.executor.v1.Value.newBuilder();
-      if (result.isNull(i)) {
+      if (struct.isNull(i)) {
         value.setIsNull(true);
       } else {
         switch (columnType.getCode()) {
           case BOOL:
-            value.setBoolValue(result.getBoolean(i));
+            value.setBoolValue(struct.getBoolean(i));
             break;
           case FLOAT32:
-            value.setDoubleValue((double) result.getFloat(i));
+            value.setDoubleValue((double) struct.getFloat(i));
             break;
           case FLOAT64:
-            value.setDoubleValue(result.getDouble(i));
+            value.setDoubleValue(struct.getDouble(i));
             break;
           case INT64:
-            value.setIntValue(result.getLong(i));
+            value.setIntValue(struct.getLong(i));
             break;
           case STRING:
-            value.setStringValue(result.getString(i));
+            value.setStringValue(struct.getString(i));
             break;
           case BYTES:
-            value.setBytesValue(toByteString(result.getBytes(i)));
+            value.setBytesValue(toByteString(struct.getBytes(i)));
             break;
           case TIMESTAMP:
-            value.setTimestampValue(timestampToProto(result.getTimestamp(i)));
+            value.setTimestampValue(timestampToProto(struct.getTimestamp(i)));
             break;
           case DATE:
-            value.setDateDaysValue(daysFromDate(result.getDate(i)));
+            value.setDateDaysValue(daysFromDate(struct.getDate(i)));
             break;
           case NUMERIC:
-            String ascii = result.getBigDecimal(i).toPlainString();
+            String ascii = struct.getBigDecimal(i).toPlainString();
             value.setStringValue(ascii);
             break;
           case JSON:
-            value.setStringValue(result.getJson(i));
+            value.setStringValue(struct.getJson(i));
             break;
```

Comment:
Do we need to handle the case of STRUCT (non-array)?

Reply:
A non-array STRUCT cannot be returned in a result set.

https://cloud.google.com/spanner/docs/reference/standard-sql/query-syntax#using_structs_with_select